### PR TITLE
Renovation: EventsMixin and EventsStrategy refactor

### PR DESCRIPTION
**Pull request by LazyLahtak**
Remove `Events Mixin`, refactor `Events Strategy` and all components which use `Events Mixin`.

### CODE REVIEW

File: js/core/events_strategy.js [hunk 1]
```diff
@@ -1,30 +1,40 @@
...
         if(callbacks) {
             return callbacks.has();
         }
         return false;
-    },
+    }
 
-    fireEvent: function(eventName, eventArgs) {
+    fireEvent(eventName, eventArgs) {
         var callbacks = this._events[eventName];
```
Comment (MaximKudriavtsev) [inline js/core/events_strategy.js]: Can we use es6 variable types?

`var` -> `const`

File: js/core/events_strategy.js [hunk 2]
```diff
@@ -1,55 +1,65 @@
...
         }
-    },
+    }
 
-    dispose: function() {
+    dispose() {
         each(this._events, function() {
             this.empty();
         });
     }
```
Comment (MaximKudriavtsev) [inline js/core/events_strategy.js]:
```js
each(this._events, () => {
  this.empty();
});
```

File: js/data/abstract_store.js [hunk 3]
```diff
@@ -309,21 +310,61 @@ var Store = Class.inherit({
...
+    /**
+     * @name StoreMethods.off
+     * @publicName off(eventName, eventHandler)
+     * @param1 eventName:string
+     * @param2 eventHandler:function
+     * @return this
+     */
+    off: function(eventName, eventHandler) {
+        this._eventsStrategy.off(eventName, eventHandler);
+        return this;
```
Comment (MaximKudriavtsev) [inline js/data/abstract_store.js]: Seems like we should move these methods into EventsStrategy.

And use here something like that

```js
off: function(eventName, eventHandler) {
  this._eventsStrategy.offMethod2(eventName, eventHandler);
  return this;
}
```

File: testing/tests/DevExpress.core/component.tests.js [hunk 4]
```diff
@@ -1442,12 +1442,12 @@ QUnit.module("event API", {
         component.on("event", h1);
         component.on("event", h2);
 
-        component.fireEvent("event");
+        component._eventsStrategy.fireEvent("event");
```
Comment (MaximKudriavtsev) [inline testing/tests/DevExpress.core/component.tests.js:1445]: Should we change the previous classes' API? I mean, we don't change tests exclude events files.

Comment (LazyLahtak) [reply testing/tests/DevExpress.core/component.tests.js:1445]: The `fireEvent` was a private method, but in some tests it is checked.

Comment (LazyLahtak) [reply testing/tests/DevExpress.core/component.tests.js:1445]: We decided to remove it from classes and call inside. Should we back it?

Comment (churkin) [reply testing/tests/DevExpress.core/component.tests.js:1445]: I think it is ok to get rid of `fireEvent` method. But events mixin is a public class. Is it braking change? @NickMitrokhin

File: js/ui/validation_engine.js [hunk 5]
```diff
@@ -751,8 +752,22 @@ const GroupConfig = Class.inherit({
         });
         this._pendingValidators = [];
         this._resetValidationInfo();
-    }
-}).include(EventsMixin);
+    },
+
+    on: function(eventName, eventHandler) {
```
Comment (churkin) [inline js/ui/validation_engine.js]: `on(eventName, eventHandler) {`

File: js/ui/validation_engine.js [hunk 6]
```diff
@@ -751,8 +752,22 @@ const GroupConfig = Class.inherit({
...
+        if(isPlainObject(eventName)) {
+            each(eventName, (function(e, h) { this.on(e, h); }).bind(this));
```
Comment (churkin) [inline js/ui/validation_engine.js]: use arrow functions

File: js/ui/pivot_grid/data_source.js [hunk 7]
```diff
@@ -1723,6 +1723,47 @@ module.exports = Class.inherit((function() {
...
+         * @return this
+         */
+        /**
+         * @name PivotGridDataSourceMethods.on
+         * @publicName on(events)
+         * @param1 events:object
+         * @return this
+         */
+        on: function(eventName, eventHandler) {
+            if(isPlainObject(eventName)) {
```
Comment (churkin) [inline js/ui/pivot_grid/data_source.js]: Should we move this logic to the event strategy?

File: js/core/events_strategy.js (hunk 2 above)
Comment (churkin) [reply js/core/events_strategy.js]:
```javascript
each(this._events, event => event.empty());
``` 
:)

File: js/core/events_strategy.js [hunk 8]
```diff
@@ -1,55 +1,65 @@
...
-    ctor: function(owner) {
+export class EventsStrategy {
+    constructor(owner) {
         this._events = {};
         this._owner = owner;
-    },
+    }
+    static setEventsStrategy(owner, strategy) {
+        if(strategy) {
+            if(typeof strategy === "function") {
```
Comment (churkin) [inline js/core/events_strategy.js]:
```javascript
return isFunction(strategy) ? strategy(owner) : strategy;
```

File: js/core/events_strategy.js [hunk 9]
```diff
@@ -1,55 +1,65 @@
...
+        } else {
+            return new EventsStrategy(owner);
+        }
+    }
 
-    hasEvent: function(eventName) {
-        var callbacks = this._events[eventName];
+    hasEvent(eventName) {
+        const callbacks = this._events[eventName];
         if(callbacks) {
```
Comment (churkin) [inline js/core/events_strategy.js]:
```javascript
return callbacks ? callbacks.has() : false;
```

**Review by churkin — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.core/component.tests.js (hunk 4 above)
Comment (vconst) [reply testing/tests/DevExpress.core/component.tests.js:1445]: The fireEvent method is not public, but some users use it. For example, https://supportcenter.devexpress.com/ticket/details/T656964/customstore-how-to-trigger-the-modified-event

Comment (churkin) [reply testing/tests/DevExpress.core/component.tests.js:1445]: It is not legal

**Review by MaximKudriavtsev — APPROVED** [review]

File: testing/tests/DevExpress.core/component.tests.js (hunk 4 above)
Comment (NickMitrokhin) [reply testing/tests/DevExpress.core/component.tests.js:1445]: This is an undocumented method. So, we can change it according to our needs without any notifications.

**Review by churkin — APPROVED** [review]

**Comment by churkin** [comment]
@nightskylark ping

**Review by nightskylark — COMMENTED** [review]
Maybe it's better to rename `EventsStrategy` to avoid confusion with the [DOM events](https://github.com/DevExpress/DevExtreme/blob/20_1/js/events.js). As I can see in fact it's a callbacks object not events.

**Comment by LazyLahtak** [comment]
> Maybe it's better to rename `EventsStrategy` to avoid confusion with the [DOM events](https://github.com/DevExpress/DevExtreme/blob/20_1/js/events.js). As I can see in fact it's a callbacks object not events.

`CallbacksStrategy` or `CallbacksManager`?
Should we also rename methods? (`fireEvent` -> `fireCallback` or `fire`) (same for `hasEvent`)

**Comment by nightskylark** [comment]
> > Maybe it's better to rename `EventsStrategy` to avoid confusion with the [DOM events](https://github.com/DevExpress/DevExtreme/blob/20_1/js/events.js). As I can see in fact it's a callbacks object not events.
> 
> `CallbacksStrategy` or `CallbacksManager`?
> Should we also rename methods? (`fireEvent` -> `fireCallback` or `fire`) (same for `hasEvent`)

It's just thoughts not a change request. I'm not crearly understand what is this module for so I'm not sure. As I can see these events are linked with owner instance. Maybe it could be something like `InstanceEvents` or `InstanceCallbacks`. Ofcourse in case we deside to get rid of `event` word we should do it everywhere.

**Comment by nightskylark** [comment]
Or maybe `ComponentXxx`.

**Comment by LazyLahtak** [comment]
> > > Maybe it's better to rename `EventsStrategy` to avoid confusion with the [DOM events](https://github.com/DevExpress/DevExtreme/blob/20_1/js/events.js). As I can see in fact it's a callbacks object not events.
> > 
> > 
> > `CallbacksStrategy` or `CallbacksManager`?
> > Should we also rename methods? (`fireEvent` -> `fireCallback` or `fire`) (same for `hasEvent`)
> 
> It's just thoughts not a change request. I'm not crearly understand what is this module for so I'm not sure. As I can see these events are linked with owner instance. Maybe it could be something like `InstanceEvents` or `InstanceCallbacks`. Ofcourse in case we deside to get rid of `event` word we should do it everywhere.

This is old `EventStrategy`. I just got rid of `EventMixin` and replace functional into components and `EventStrategy`. `EventStrategy` contains callbacks (events) and fire it (for example `disposing` or `optionChanged`)

**Comment by LazyLahtak** [comment]
@nightskylark,

As I see, with rename of the`eventsStrategy` we have to change (rename) some elements. I think we should do it in other PR.

**Review by DokaRus — APPROVED** [review]

**Pull request merged**